File: desktop/src/test/unit/sync/LocatorCodec.golden.test.ts
import { describe, expect, it } from 'vitest';
import {
  locatorFromCfi,
  locatorFromJson,
  locatorToJson,
  normalizeHref,
  normalizeLocatorJson,
  parseSpineIndex,
} from '$lib/shared/sync/LocatorCodec';

// Golden vectors shared with Android LocatorCodec.kt
// These vectors MUST stay in sync cross-language.
// If you update them here, update android/app/src/test/java/com/nextpage/data/sync/LocatorCodecGoldenTest.kt
const GOLDEN_VECTORS = [
  {
    name: 'backslash href normalized via normalizeHref',
    input: 'OEBPS\\Text\\chapter1.xhtml',
    expected: 'OEBPS/Text/chapter1.xhtml',
  },
  {
    name: 'backslash in readingOrder resolved via spineHrefs',
    readingOrder: ['OEBPS\\Text\\chap1.xhtml', 'OEBPS/Text/chap2.xhtml', 'OEBPS/Text/chap3.xhtml'],
    cfi: 'epubcfi(/6/2!/4/2,/1:0,/1:5)',
    expectedHref: 'OEBPS/Text/chap2.xhtml',
  },
  {
    name: 'locatorJson backslash normalized',
    json: '{"href":"OEBPS\\\\Text\\\\chap1.xhtml","type":"application/xhtml+xml","locations":{"progression":0.5,"fragment":"epubcfi(/6/1!/4/2,/1:0,/1:5)"}}',
    expectedHref: 'OEBPS/Text/chap1.xhtml',
  },
] as const;

describe('LocatorCodec golden cross-language', () => {
  it('normalizeHref converts backslashes to forward slashes (parity with Kotlin)', () => {
    expect(normalizeHref('a\\b\\c')).toBe('a/b/c');
    expect(normalizeHref('OEBPS\\Text\\chap1.xhtml')).toBe('OEBPS/Text/chap1.xhtml');
    expect(normalizeHref('already/correct')).toBe('already/correct');
    expect(normalizeHref(GOLDEN_VECTORS[0].input)).toBe(GOLDEN_VECTORS[0].expected);
  });

  it('parseSpineIndex extracts spine index correctly (parity with Kotlin SPINE_INDEX_RE)', () => {
    expect(parseSpineIndex('epubcfi(/6/1!/4/2,/1:0,/1:5)')).toBe(1);
    expect(parseSpineIndex('epubcfi(/6/24!/4/2,/1:0,/1:5)')).toBe(24);
    expect(parseSpineIndex('epubcfi(/6/2!foo)')).toBe(2);
    expect(parseSpineIndex('invalid')).toBeNull();
    expect(parseSpineIndex(null as unknown as string)).toBeNull();
  });

  it('readingOrder = spineHrefs: locatorFromCfi resolves via spine index', () => {
    const readingOrder = GOLDEN_VECTORS[1].readingOrder as unknown as string[];
    const loc = locatorFromCfi(readingOrder, GOLDEN_VECTORS[1].cfi, null);
    expect(loc).not.toBeNull();
    expect(loc!.href).toBe(GOLDEN_VECTORS[1].expectedHref);
    // Also verify backslash in readingOrder is normalized
    const loc2 = locatorFromCfi(readingOrder, 'epubcfi(/6/1!/4/2,/1:0,/1:5)', null);
    expect(loc2!.href).toBe('OEBPS/Text/chap1.xhtml');
  });

  it('locatorToJson normalizes href backslashes before serialization', () => {
    const loc = locatorFromCfi(['OEBPS\\Text\\chap1.xhtml', 'b.html'], 'epubcfi(/6/1!/4/2,/1:0,/1:5)', null)!;
    const json = locatorToJson(loc);
    const parsed = JSON.parse(json) as { href: string };
    expect(parsed.href).toBe('OEBPS/Text/chap1.xhtml');
    expect(json).not.toContain('\\');
  });

  it('locatorFromJson normalizes backslash href on deserialization (golden)', () => {
    const v = GOLDEN_VECTORS[2];
    const loc = locatorFromJson(v.json as string);
    expect(loc).not.toBeNull();
    expect(loc!.href).toBe(v.expectedHref);
  });

  it('normalizeLocatorJson parity: backslash in href replaced, CFI preserved', () => {
    const json = GOLDEN_VECTORS[2].json as string;
    const normalized = normalizeLocatorJson(json)!;
    const parsed = JSON.parse(normalized) as { href: string; locations: { fragment: string } };
    expect(parsed.href).toBe(GOLDEN_VECTORS[2].expectedHref);
    expect(parsed.locations.fragment).toBe('epubcfi(/6/1!/4/2,/1:0,/1:5)');
  });

  it('round-trip spineHrefs golden: Historia offset-2 simulation', () => {
    // Simulate Historia: spine 4, TOC 0 offset (spine[2] is first TOC entry)
    const spineHrefs = [
      'OEBPS/Text/cover.xhtml',
      'OEBPS/Text/toc.xhtml',
      'OEBPS/Text/HM-colombia-1.html',
      'OEBPS/Text/HM-colombia-2.html',
      'OEBPS/Text/HM-colombia-3.html',
    ];
    // Toc entry 0 maps to spine index 2 (0-based) => cfi /6/3
    const cfiForToc0 = 'epubcfi(/6/3!/4/2,/1:0,/1:5)';
    const loc = locatorFromCfi(spineHrefs, cfiForToc0, { chapterChars: 1000, charOffset: 250 });
    expect(loc).not.toBeNull();
    expect(loc!.href).toBe('OEBPS/Text/HM-colombia-1.html');
    expect(loc!.locations.progression).toBeCloseTo(0.25, 5);
    expect(loc!.locations.fragment).toBe(cfiForToc0);
    const json = locatorToJson(loc!);
    const restored = locatorFromJson(json)!;
    expect(restored.href).toBe(loc!.href);
    expect(restored.locations.fragment).toBe(cfiForToc0);
  });
});

File: android/app/src/test/java/com/nextpage/data/sync/LocatorCodecGoldenTest.kt
package com.nextpage.data.sync

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Cross-language golden test — parity with desktop LocatorCodec.ts
 * Golden vectors are defined in desktop/src/test/unit/sync/LocatorCodec.golden.test.ts
 * This test MUST stay in sync with the TS golden. If you update one, update the other.
 *
 * Covers:
 * - readingOrder = spineHrefs (authoritative OPF spine order, linear=no filtered)
 * - normalizeHref parity (backslash -> forward slash)
 * - locatorFromCfi / locatorToJson / locatorFromJson round-trip
 */
class LocatorCodecGoldenTest {

    @Test
    fun normalizeHref_convertsBackslashes_parityWithTs() {
        assertEquals("a/b/c", LocatorCodec.normalizeHref("a\\b\\c"))
        assertEquals("OEBPS/Text/chapter1.xhtml", LocatorCodec.normalizeHref("OEBPS\\Text\\chapter1.xhtml"))
        assertEquals("already/correct", LocatorCodec.normalizeHref("already/correct"))
        // Golden vector 0
        assertEquals("OEBPS/Text/chapter1.xhtml", LocatorCodec.normalizeHref("OEBPS\\Text\\chapter1.xhtml"))
    }

    @Test
    fun parseSpineIndex_parityWithTs() {
        assertEquals(1, LocatorCodec.parseSpineIndex("epubcfi(/6/1!/4/2,/1:0,/1:5)"))
        assertEquals(24, LocatorCodec.parseSpineIndex("epubcfi(/6/24!/4/2,/1:0,/1:5)"))
        assertEquals(2, LocatorCodec.parseSpineIndex("epubcfi(/6/2!foo)"))
        assertNull(LocatorCodec.parseSpineIndex("invalid"))
        assertNull(LocatorCodec.parseSpineIndex(null))
    }

    @Test
    fun readingOrder_spineHrefs_locatorFromCfi_resolvesViaSpineIndex() {
        val readingOrder = listOf("OEBPS\\Text\\chap1.xhtml", "OEBPS/Text/chap2.xhtml", "OEBPS/Text/chap3.xhtml")
        val loc = LocatorCodec.locatorFromCfi(readingOrder, "epubcfi(/6/2!/4/2,/1:0,/1:5)", null)
        assertNotNull(loc)
        assertEquals("OEBPS/Text/chap2.xhtml", loc!!.href)
        // Backslash in readingOrder normalized
        val loc2 = LocatorCodec.locatorFromCfi(readingOrder, "epubcfi(/6/1!/4/2,/1:0,/1:5)", null)
        assertNotNull(loc2)
        assertEquals("OEBPS/Text/chap1.xhtml", loc2!!.href)
    }

    @Test
    fun locatorToJson_normalizesHrefBeforeSerialization() {
        val loc = LocatorCodec.locatorFromCfi(listOf("OEBPS\\Text\\chap1.xhtml", "b.html"), "epubcfi(/6/1!/4/2,/1:0,/1:5)", null)!!
        val json = LocatorCodec.locatorToJson(loc)
        assertTrue(!json.contains("\\"))
        val parsed = org.json.JSONObject(json)
        assertEquals("OEBPS/Text/chap1.xhtml", parsed.getString("href"))
    }

    @Test
    fun locatorFromJson_normalizesBackslashHref_golden() {
        val json = """{"href":"OEBPS\\Text\\chap1.xhtml","type":"application/xhtml+xml","locations":{"progression":0.5,"fragment":"epubcfi(/6/1!/4/2,/1:0,/1:5)"}}"""
        val loc = LocatorCodec.locatorFromJson(json)
        assertNotNull(loc)
        assertEquals("OEBPS/Text/chap1.xhtml", loc!!.href)
    }

    @Test
    fun normalizeLocatorJson_parity_backslashInHrefReplaced_cfiPreserved() {
        val json = """{"href":"OEBPS\\Text\\chap1.xhtml","type":"application/xhtml+xml","locations":{"progression":0.5,"fragment":"epubcfi(/6/1!/4/2,/1:0,/1:5)"}}"""
        val normalized = LocatorCodec.normalizeLocatorJson(json)
        assertNotNull(normalized)
        val parsed = org.json.JSONObject(normalized!!)
        assertEquals("OEBPS/Text/chap1.xhtml", parsed.getString("href"))
        val loc = parsed.getJSONObject("locations")
        assertEquals("epubcfi(/6/1!/4/2,/1:0,/1:5)", loc.getString("fragment"))
    }

    @Test
    fun roundTrip_spineHrefs_golden_HistoriaOffset2() {
        val spineHrefs = listOf(
            "OEBPS/Text/cover.xhtml",
            "OEBPS/Text/toc.xhtml",
            "OEBPS/Text/HM-colombia-1.html",
            "OEBPS/Text/HM-colombia-2.html",
            "OEBPS/Text/HM-colombia-3.html"
        )
        val cfiForToc0 = "epubcfi(/6/3!/4/2,/1:0,/1:5)"
        val loc = LocatorCodec.locatorFromCfi(spineHrefs, cfiForToc0, LocatorChapterMetric(chapterChars = 1000, charOffset = 250))
        assertNotNull(loc)
        assertEquals("OEBPS/Text/HM-colombia-1.html", loc!!.href)
        assertEquals(0.25, loc.locations.progression!!, 0.00001)
        assertEquals(cfiForToc0, loc.locations.fragment)
        val json = LocatorCodec.locatorToJson(loc)
        val restored = LocatorCodec.locatorFromJson(json)!!
        assertEquals(loc.href, restored.href)
        assertEquals(cfiForToc0, restored.locations.fragment)
    }

    @Test
    fun charOffsetToProgression_clamped() {
        assertEquals(0.25, LocatorCodec.charOffsetToProgression(250, 1000)!!, 0.00001)
        assertEquals(1.0, LocatorCodec.charOffsetToProgression(1500, 1000)!!, 0.00001)
        assertEquals(0.0, LocatorCodec.charOffsetToProgression(-5, 100)!!, 0.00001)
        assertNull(LocatorCodec.charOffsetToProgression(10, 0))
    }

    @Test
    fun deriveLocatorForChapter_parity() {
        val readingOrder = listOf("a.html", "b.html", "c.html")
        val loc = LocatorCodec.deriveLocatorForChapter(readingOrder, "b.html")
        assertNotNull(loc)
        assertEquals("b.html", loc!!.href)
        assertEquals(0.0, loc.locations.progression!!, 0.00001)
        assertNull(LocatorCodec.deriveLocatorForChapter(readingOrder, "missing.html"))
        // Backslash normalized
        val loc2 = LocatorCodec.deriveLocatorForChapter(listOf("OEBPS\\Text\\a.html", "b.html"), "OEBPS/Text/a.html")
        assertNotNull(loc2)
    }

    @Test
    fun cfiFromLocator_roundTrip() {
        val loc = LocatorCodec.locatorFromCfi(listOf("a.html", "b.html"), "epubcfi(/6/2!/4/2,/1:0,/1:5)", null)!!
        assertEquals("epubcfi(/6/2!/4/2,/1:0,/1:5)", LocatorCodec.cfiFromLocator(loc))
        val chapLoc = LocatorCodec.deriveLocatorForChapter(listOf("a.html", "b.html"), "a.html")!!
        assertNull(LocatorCodec.cfiFromLocator(chapLoc))
    }
}
